File: paypay/src/main/scala/jp/ijufumi/practices/paypay/scala/impl/QueueImpl.scala
package jp.ijufumi.practices.paypay.scala.impl

import jp.ijufumi.practices.paypay.scala.Queue

import scala.reflect.ClassTag

class QueueImpl[T: ClassTag](val elements: Array[T]) extends Queue[T] {
  override def isEmpty: Boolean = elements.length == 0

  override def enQueue(t: T): Queue[T] = {
    val newElements:Array[T] = elements :+ t
    new QueueImpl[T](newElements)
  }

  override def deQueue(): Queue[T] = {
    if (elements.isEmpty) {
      return Queue.empty[T]
    }
    val newElements = elements.slice(1, elements.length)
    new QueueImpl[T](newElements)
  }

  override def head: Option[T] = elements.headOption
}

File: paypay/src/main/java/jp/ijufumi/practices/paypay/impl/QueueImpl.java
package jp.ijufumi.practices.paypay.impl;

import jp.ijufumi.practices.paypay.Queue;

import java.util.Arrays;

public class QueueImpl<T> implements Queue<T> {

    private Object[] elements;

    public QueueImpl() {
        this(new Object[0]);
    }

    public QueueImpl(Object..._elements) {
        if (_elements == null) {
            throw new IllegalArgumentException("elements not allowed null.");
        }
        elements = _elements;
    }

    public Queue<T> enQueue(T t) {
        var newElements = new Object[elements.length + 1];
        System.arraycopy(elements, 0, newElements, 0, elements.length);
        newElements[elements.length] = t;
        return new QueueImpl<>(newElements);
    }

    public Queue<T> deQueue() {
        var newElements = new Object[elements.length - 1];
        System.arraycopy(elements, 1, newElements, 0, elements.length - 1);
        return new QueueImpl<>(newElements);
    }

    @SuppressWarnings("unchecked")
    public T head() {
        return (T) Arrays.stream(elements).findFirst().orElse(null);
    }

    public boolean isEmpty() {
        return elements.length == 0;
    }
}
